File: src/services/NeoService.ts
/**
 * NEO Blockchain Service - Frontend
 * Mock mode for MVP - no backend/blockchain required
 * All data stored in React state (resets on page reload)
 */

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000';

// ============================================
// MOCK MODE CONFIGURATION
// ============================================
const USE_MOCK_MODE = true; // Set to false to use real backend

export interface NeoBlockchainStatus {
  network: string;
  rpc: string;
  blockHeight: number;
  status: string;
  walletAddress?: string;
  balance?: {
    neo: string;
    gas: string;
  };
}

export interface NeOMission {
  mission_id: number;
  creator: string;
  title: string;
  description: string;
  reward_amount: number;
  deadline: number;
  category: string;
  status: number; // 0=PENDING, 1=IN_PROGRESS, 2=COMPLETED, 3=VERIFIED, 4=CANCELLED
  assigned_user?: string;
  created_at: number;
  completed_at?: number;
  verified_at?: number;
  proof?: string;
}

export interface CreateMissionRequest {
  creator: string;
  title: string;
  description: string;
  reward_amount: number;
  deadline: number;
  category: string;
}

export interface MissionActionResult {
  success: boolean;
  txHash?: string;
  message?: string;
  error?: string;
}

export class NeoService {
  private walletAddress: string = 'NWWkFU3dKWTHNpxjz8MRgt5eKe1Ld834xQ'; // Default mock wallet
  private mockMissionCounter: number = 100; // For generating IDs

  /**
   * Set current user's NEO wallet address
   */
  setWalletAddress(address: string) {
    this.walletAddress = address;
  }

  /**
   * Get current wallet address
   */
  getWalletAddress(): string | null {
    return this.walletAddress;
  }

  /**
   * Generate mock transaction hash
   */
  private generateMockTxHash(): string {
    return `0x${Date.now().toString(16)}${Math.random().toString(16).slice(2, 10)}`;
  }

  /**
   * Get NEO blockchain status
   */
  async getBlockchainStatus(): Promise<NeoBlockchainStatus> {
    if (USE_MOCK_MODE) {
      // Mock response
      await new Promise(resolve => setTimeout(resolve, 300)); // Simulate network delay
      return {
        network: 'NEO N3 TestNet',
        rpc: 'http://seed1t5.neo.org:20332',
        blockHeight: 11108000 + Math.floor(Math.random() * 1000),
        status: 'connected',
        walletAddress: this.walletAddress,
        balance: {
          neo: '0.0000005',
          gas: '50.0'
        }
      };
    }

    try {
      const response = await fetch(`${API_BASE_URL}/api/neo/status`);
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }
      return await response.json();
    } catch (error: any) {
      console.error('Failed to get NEO status:', error);
      throw new Error(error.message || 'Failed to connect to NEO blockchain');
    }
  }

  /**
   * Create a new mission on NEO blockchain
   */
  async createMission(missionData: CreateMissionRequest): Promise<{
    success: boolean;
    missionId?: number;
    txHash?: string;
    error?: string;
  }> {
    if (USE_MOCK_MODE) {
      // Mock mode: simulate blockchain transaction
      await new Promise(resolve => setTimeout(resolve, 500));
      
      const missionId = ++this.mockMissionCounter;
      const txHash = this.generateMockTxHash();
      
      console.log('✅ Mock mission created:', { missionId, txHash, missionData });
      
      return {
        success: true,
        missionId,
        txHash,
      };
    }

    try {
      const response = await fetch(`${API_BASE_URL}/api/neo/create-mission`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(missionData),
      });

      const result = await response.json();

      if (!response.ok) {
        return {
          success: false,
          error: result.detail || 'Failed to create mission',
        };
      }

      return {
        success: true,
        missionId: result.missionId,
        txHash: result.txHash,
      };
    } catch (error: any) {
      console.error('Create mission error:', error);
      return {
        success: false,
        error: error.message || 'Failed to create mission',
      };
    }
  }

  /**
   * Get mission details
   */
  async getMission(missionId: number): Promise<NeOMission | null> {
    if (USE_MOCK_MODE) {
      await new Promise(resolve => setTimeout(resolve, 200));
      
      // Mock mission data
      return {
        mission_id: missionId,
        creator: this.walletAddress,
        title: `Mission #${missionId}`,
        description: 'Mock mission description',
        reward_amount: 10,
        deadline: Date.now() + 86400000,
        category: 'fitness',
        status: 0, // PENDING
        assigned_user: '',
        created_at: Date.now() - 3600000,
      };
    }

    try {
      const response = await fetch(`${API_BASE_URL}/api/missions/${missionId}`);
      
      if (!response.ok) {
        if (response.status === 404) {
          return null;
        }
        throw new Error(`HTTP ${response.status}`);
      }

      return await response.json();
    } catch (error: any) {
      console.error('Get mission error:', error);
      return null;
    }
  }

  /**
   * Accept a mission
   */
  async acceptMission(missionId: number, userAddress?: string): Promise<MissionActionResult> {
    if (USE_MOCK_MODE) {
      await new Promise(resolve => setTimeout(resolve, 400));
      
      const txHash = this.generateMockTxHash();
      console.log('✅ Mock mission accepted:', { missionId, txHash });
      
      return {
        success: true,
        txHash,
        message: 'Mission accepted successfully',
      };
    }

    try {
      const user = userAddress || this.walletAddress;
      
      if (!user) {
        return {
          success: false,
          error: 'Wallet address not set. Please connect your NEO wallet first.',
        };
      }

      const response = await fetch(`${API_BASE_URL}/api/missions/${missionId}/accept`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ user }),
      });

      const result = await response.json();

      if (!response.ok) {
        return {
          success: false,
          error: result.detail || 'Failed to accept mission',
        };
      }

      return {
        success: result.success,
        txHash: result.txHash,
        message: result.message,
      };
    } catch (error: any) {
      console.error('Accept mission error:', error);
      return {
        success: false,
        error: error.message || 'Failed to accept mission',
      };
    }
  }

  /**
   * Complete a mission with proof
   */
  async completeMission(
    missionId: number,
    proof: string,
    userAddress?: string
  ): Promise<MissionActionResult> {
    if (USE_MOCK_MODE) {
      await new Promise(resolve => setTimeout(resolve, 600));
      
      const txHash = this.generateMockTxHash();
      console.log('✅ Mock mission completed:', { missionId, proof, txHash });
      
      return {
        success: true,
        txHash,
        message: 'Mission completed, awaiting verification',
      };
    }

    try {
      const user = userAddress || this.walletAddress;
      
      if (!user) {
        return {
          success: false,
          error: 'Wallet address not set',
        };
      }

      const response = await fetch(`${API_BASE_URL}/api/missions/${missionId}/complete`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ user, proof }),
      });

      const result = await response.json();

      if (!response.ok) {
        return {
          success: false,
          error: result.detail || 'Failed to complete mission',
        };
      }

      return {
        success: result.success,
        txHash: result.txHash,
        message: result.message,
      };
    } catch (error: any) {
      console.error('Complete mission error:', error);
      return {
        success: false,
        error: error.message || 'Failed to complete mission',
      };
    }
  }

  /**
   * Verify a completed mission
   */
  async verifyMission(
    missionId: number,
    approved: boolean,
    verifierAddress?: string
  ): Promise<MissionActionResult> {
    try {
      const verifier = verifierAddress || this.walletAddress;
      
      if (!verifier) {
        return {
          success: false,
          error: 'Wallet address not set',
        };
      }

      const response = await fetch(`${API_BASE_URL}/api/missions/${missionId}/verify`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ verifier, approved }),
      });

      const result = await response.json();

      if (!response.ok) {
        return {
          success: false,
          error: result.detail || 'Failed to verify mission',
        };
      }

      return {
        success: result.success,
        txHash: result.txHash,
        message: result.message,
      };
    } catch (error: any) {
      console.error('Verify mission error:', error);
      return {
        success: false,
        error: error.message || 'Failed to verify mission',
      };
    }
  }

  /**
   * Get all missions for a user
   */
  async getUserMissions(userAddress?: string): Promise<{
    missionIds: number[];
    missions: NeOMission[];
    count: number;
  } | null> {
    try {
      const user = userAddress || this.walletAddress;
      
      if (!user) {
        throw new Error('Wallet address not set');
      }

      const response = await fetch(`${API_BASE_URL}/api/missions/user/${user}`);
      
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }

      return await response.json();
    } catch (error: any) {
      console.error('Get user missions error:', error);
      return null;
    }
  }

  /**
   * Get total mission count
   */
  async getMissionCount(): Promise<number> {
    try {
      const response = await fetch(`${API_BASE_URL}/api/missions/count`);
      
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }

      const result = await response.json();
      return result.count || 0;
    } catch (error: any) {
      console.error('Get mission count error:', error);
      return 0;
    }
  }

  /**
   * Check if service is healthy
   */
  async healthCheck(): Promise<boolean> {
    try {
      const response = await fetch(`${API_BASE_URL}/health`);
      return response.ok;
    } catch (error) {
      console.error('Health check failed:', error);
      return false;
    }
  }
}

// Export singleton instance
export const neoService = new NeoService();

File: spoonos-backend/neo_service.py
"""
NEO Blockchain Service
Service layer để tương tác với NEO smart contracts
"""
import os
import json
import requests
from typing import Dict, Any, Optional, List
from dotenv import load_dotenv

load_dotenv()


class NeoService:
    """
    Service để tương tác với NEO blockchain
    """
    
    def __init__(self):
        self.rpc_url = os.getenv("NEO_RPC_URL", "http://seed1t5.neo.org:20332")
        self.wallet_address = os.getenv("NEO_WALLET_ADDRESS", "")
        self.private_key = os.getenv("NEO_WALLET_PRIVATE_KEY", "")
        self.mission_contract = os.getenv("NEO_MISSION_CONTRACT", "")
        
    def _rpc_call(self, method: str, params: List = None) -> Dict[str, Any]:
        """
        Make RPC call to NEO node
        
        Args:
            method: RPC method name
            params: Method parameters
            
        Returns:
            RPC response data
        """
        if params is None:
            params = []
            
        payload = {
            "jsonrpc": "2.0",
            "method": method,
            "params": params,
            "id": 1
        }
        
        try:
            response = requests.post(
                self.rpc_url,
                json=payload,
                headers={"Content-Type": "application/json"},
                timeout=30
            )
            
            if response.status_code == 200:
                data = response.json()
                if "result" in data:
                    return data["result"]
                elif "error" in data:
                    raise Exception(f"RPC Error: {data['error']}")
            
            raise Exception(f"HTTP Error: {response.status_code}")
            
        except Exception as e:
            raise Exception(f"RPC call failed: {e}")
    
    def get_block_count(self) -> int:
        """Get current block height"""
        return self._rpc_call("getblockcount")
    
    def get_balance(self, address: str = None) -> Dict[str, float]:
        """
        Get GAS and NEO balance
        
        Args:
            address: NEO address (default: wallet address)
            
        Returns:
            Dict with 'GAS' and 'NEO' balances
        """
        if not address:
            address = self.wallet_address
            
        result = self._rpc_call("getnep17balances", [address])
        
        balances = {"GAS": 0.0, "NEO": 0.0}
        
        if "balance" in result:
            for balance in result["balance"]:
                asset_hash = balance.get("assethash", "")
                amount = int(balance.get("amount", 0)) / 100000000
                
                # GAS token
                if asset_hash == "0xd2a4cff31913016155e38e474a2c06d08be276cf":
                    balances["GAS"] = amount
                # NEO token
                elif asset_hash == "0xef4073a0f2b305a38ec4050e4d3d28bc40ea63f5":
                    balances["NEO"] = amount
        
        return balances
    
    def invoke_contract_read(self, contract_hash: str, operation: str, params: List = None) -> Any:
        """
        Invoke contract method (read-only)
        
        Args:
            contract_hash: Contract hash
            operation: Method name
            params: Method parameters
            
        Returns:
            Contract return value
        """
        if params is None:
            params = []
        
        invoke_params = [
            contract_hash,
            operation,
            params
        ]
        
        result = self._rpc_call("invokefunction", invoke_params)
        
        if "stack" in result and len(result["stack"]) > 0:
            return result["stack"][0]
        
        return None
    
    # ========== Mission Contract Methods ==========
    
    def create_mission(
        self,
        creator: str,
        title: str,
        description: str,
        reward_amount: int,
        deadline: int,
        category: str
    ) -> Dict[str, Any]:
        """
        Create new mission on blockchain
        
        Args:
            creator: Creator address
            title: Mission title
            description: Mission description
            reward_amount: Reward in GAS
            deadline: Unix timestamp
            category: Mission category
            
        Returns:
            Transaction info with mission_id
        """
        # For now, return mock data until contract is deployed
        # TODO: Replace with actual contract invocation
        
        mission_id = 1  # Mock ID
        
        return {
            "success": True,
            "mission_id": mission_id,
            "tx_hash": "0xmock_transaction_hash",
            "message": "Mission created successfully (mock)",
            "data": {
                "id": mission_id,
                "creator": creator,
                "title": title,
                "description": description,
                "reward": reward_amount,
                "deadline": deadline,
                "category": category,
                "status": "PENDING"
            }
        }
    
    def get_mission(self, mission_id: int) -> Optional[Dict[str, Any]]:
        """
        Get mission details from blockchain
        
        Args:
            mission_id: Mission ID
            
        Returns:
            Mission data or None
        """
        # Mock data until contract deployed
        # TODO: Replace with actual contract call
        
        return {
            "id": mission_id,
            "creator": self.wallet_address,
            "title": f"Mission #{mission_id}",
            "description": "Complete this awesome mission!",
            "reward": 5,
            "deadline": 1700000000,
            "category": "fitness",
            "status": "PENDING",
            "assignee": "",
            "created_at": 1699000000,
            "completed_at": 0
        }
    
    def accept_mission(self, mission_id: int, user: str) -> Dict[str, Any]:
        """
        User accepts a mission
        
        Args:
            mission_id: Mission ID
            user: User address
            
        Returns:
            Transaction result
        """
        # Mock response
        return {
            "success": True,
            "tx_hash": "0xmock_accept_tx",
            "message": f"Mission {mission_id} accepted by {user}",
            "data": {
                "mission_id": mission_id,
                "status": "IN_PROGRESS",
                "assignee": user
            }
        }
    
    def complete_mission(self, mission_id: int, user: str, proof: str) -> Dict[str, Any]:
        """
        Mark mission as completed
        
        Args:
            mission_id: Mission ID
            user: User address
            proof: Proof URL/hash
            
        Returns:
            Transaction result
        """
        return {
            "success": True,
            "tx_hash": "0xmock_complete_tx",
            "message": f"Mission {mission_id} completed",
            "data": {
                "mission_id": mission_id,
                "status": "COMPLETED",
                "proof": proof
            }
        }
    
    def verify_mission(self, mission_id: int, verifier: str, approved: bool) -> Dict[str, Any]:
        """
        Verify completed mission
        
        Args:
            mission_id: Mission ID
            verifier: Verifier address
            approved: True to approve, False to reject
            
        Returns:
            Transaction result
        """
        return {
            "success": True,
            "tx_hash": "0xmock_verify_tx",
            "message": f"Mission {mission_id} {'approved' if approved else 'rejected'}",
            "data": {
                "mission_id": mission_id,
                "status": "VERIFIED" if approved else "IN_PROGRESS",
                "reward_distributed": approved
            }
        }
    
    def get_user_missions(self, user: str) -> List[int]:
        """
        Get list of user's missions
        
        Args:
            user: User address
            
        Returns:
            List of mission IDs
        """
        # Mock data
        return [1, 2, 3, 5, 7]
    
    def get_mission_count(self) -> int:
        """
        Get total number of missions
        
        Returns:
            Mission count
        """
        # Mock data
        return 42
    
    def get_blockchain_status(self) -> Dict[str, Any]:
        """
        Get NEO blockchain status
        
        Returns:
            Status information
        """
        try:
            block_count = self.get_block_count()
            balances = self.get_balance()
            
            return {
                "connected": True,
                "network": "NEO N3 TestNet",
                "block_height": block_count,
                "wallet_address": self.wallet_address,
                "balances": balances,
                "contract_address": self.mission_contract,
                "rpc_url": self.rpc_url
            }
        except Exception as e:
            return {
                "connected": False,
                "error": str(e),
                "network": "NEO N3 TestNet",
                "rpc_url": self.rpc_url
            }


# Singleton instance
neo_service = NeoService()
